packages/kbn-securitysolution-io-ts-alerting-types/src/actions/index.ts

```diff
@@ -25,27 +25,44 @@ export const RuleActionTypeId = t.string;
 export type RuleActionParams = t.TypeOf<typeof RuleActionParams>;
 export const RuleActionParams = saved_object_attributes;
 
+export type RuleActionFrequency = t.TypeOf<typeof RuleActionFrequency>;
+export const RuleActionFrequency = t.type({
+  summary: t.boolean,
+  notifyWhen: t.union([
+    t.literal('onActionGroupChange'),
+    t.literal('onActiveAlert'),
+    t.literal('onThrottleInterval'),
+  ]),
+  throttle: t.union([t.string, t.null]),
+});
+
 export type RuleAction = t.TypeOf<typeof RuleAction>;
 export const RuleAction = t.exact(
-  t.type({
-    group: RuleActionGroup,
-    id: RuleActionId,
-    action_type_id: RuleActionTypeId,
-    params: RuleActionParams,
-  })
+  t.intersection([
+    t.type({
+      group: RuleActionGroup,
+      id: RuleActionId,
+      action_type_id: RuleActionTypeId,
+      params: RuleActionParams,
+    }),
+    t.partial({ frequency: RuleActionFrequency }),
```

Comment:
This line allows setting action frequency via API, but according to the PR description:

> The Security Solution UX remains unchanged, as it has a unique way of displaying action notification frequencies at the rule level. Instead, the API request has changed so that the selected action frequency will now be stored in each action's frequency param instead of at the rule level.

What will be displayed in the UI if users modify action frequency via API and set values not supported on the front end? I'm pretty sure there will be a discrepancy between what's displayed and the actual action values.

Reply:
Which API methods are you concerned about this happening with? I've added `syncFrequency` functionality to the security solution's bulk edit API, which forces the frequency of action `0` to be applied to all other actions. You can see this in the `rules_client.ts` changes below.

I did this to fix failing tests, so it's more difficult for me to understand which other methods I need to add this functionality to. Are you able to point them out for me, and I can make sure they don't allow discrepancies until when and if the security solution UI gets a redesign?

Reply:
Okay, actually, interesting news. I just found out that what you're asking for is already happening. 

The security solution's create, update, etc. APIs call [`transformRuleToAlertAction`](https://github.com/elastic/kibana/blob/6f3e0a3b952ce26f5e183c3bbda360813cedf3b8/x-pack/plugins/security_solution/common/detection_engine/transform_actions.ts#LL12C29-L12C29), which has the effect of removing any and all `frequency` props from actions before creating the SavedObject. So these APIs will still only actually commit the legacy rule-level `notifyWhen` or `throttle` properties, regardless of whether they're called from the UI or another context.

The bulk edit API doesn't use this transformation, so after pushing this PR, the bulk edit API will be the only security solution API that's actually capable of committing an action with a `frequency` parameter. But because of `syncFrequency` it forces all of these `frequency` parameters to be the same, just like the UI. Rules edited with this API will behave identically to rules edited with the non-bulk `update` API.

It's a bit of an uneven migration pathway for the Security Solution but hopefully this should be good enough to unblock this change for y'all.

Reply:
> The security solution's create, update, etc. APIs call [`transformRuleToAlertAction`](https://github.com/elastic/kibana/blob/6f3e0a3b952ce26f5e183c3bbda360813cedf3b8/x-pack/plugins/security_solution/common/detection_engine/transform_actions.ts#LL12C29-L12C29), which has the effect of removing any and all `frequency` props from actions before creating the SavedObject. So these APIs will still only actually commit the legacy rule-level `notifyWhen` or `throttle` properties, regardless of whether they're called from the UI or another context.

Okay, so the newly introduced `frequency` param is accepted by our CRUD APIs but ignored in API handlers. Let's remove `frequency` from the contract schema, then.

> The bulk edit API doesn't use this transformation, so after pushing this PR, the bulk edit API will be the only security solution API that's actually capable of committing an action with a `frequency` parameter. But because of `syncFrequency` it forces all of these `frequency` parameters to be the same, just like the UI. Rules edited with this API will behave identically to rules edited with the non-bulk `update` API.

I propose reverting that change to the bulk API to align capabilities throughout all APIs in the detection engine. I would prefer to have all our APIs migrated to the action-level frequency simultaneously to avoid confusion for our users (as all APIs are public) and to prevent workarounds, like setting all action frequencies to the same value.

Reply:
I may not be able to leave the API completely unchanged, because I initially did this to fix breaking Bulk Edit tests. But I'll figure something out that preserves existing functionality.